src/EnergyPlus/ZoneEquipmentManager.cc

```diff
@@ -4073,6 +3996,145 @@ namespace ZoneEquipmentManager {
 
 		} while ( Iteration < IterMax );
 
+		// Check for unbalanced airloop
+		for ( int AirLoopNum = 1; AirLoopNum <= NumPrimaryAirSys; ++AirLoopNum ) {
+			{ auto & thisAirLoopFlow ( AirLoopFlow( AirLoopNum ) );
+			thisAirLoopFlow.SysRetFlow = thisAirLoopFlow.ZoneRetFlow - thisAirLoopFlow.RecircFlow + thisAirLoopFlow.LeakFlow;
+			if ( !isPulseZoneSizing && !ZoneAirMassFlow.EnforceZoneMassBalance && !DataGlobals::WarmupFlag && AirLoopsSimOnce ) {
+				Real64 exhaustDelta = thisAirLoopFlow.SupFlow - thisAirLoopFlow.SysRetFlow;
+				Real64 unbalancedExhaustDelta = max( 0.0, ( exhaustDelta - thisAirLoopFlow.MaxOutAir ) );
+				if ( ( unbalancedExhaustDelta > SmallMassFlow) && !thisAirLoopFlow.FlowError ) {
+					ShowWarningError( "In AirLoopHVAC " + PrimaryAirSystem( AirLoopNum ).Name + " there is unbalanced exhaust air flow." );
+					ShowContinueErrorTimeStamp( "" );
+					ShowContinueError( "  Unless there is balancing infiltration / ventilation air flow, this will result in" );
+					ShowContinueError( "  load due to induced outdoor air being neglected in the simulation." );
+					thisAirLoopFlow.FlowError = true;
+				}
+			}
+		}}
+
+	}
+
+	void
+	CalcZoneReturnFlows(
+		int const ZoneNum,
+		Real64 & ExpTotalReturnMassFlow, // Expected total return air mass flow rate
+		Real64 & FinalTotalReturnMassFlow // Final total return air mass flow rate
+	)
+	{
+		auto & thisZoneEquip( ZoneEquipConfig( ZoneNum ) );
+		int numRetNodes = thisZoneEquip.NumReturnNodes;
+		Real64 totReturnFlow = 0.0; // Total flow to all return nodes in the zone (kg/s)
+		Real64 totVarReturnFlow = 0.0; // Total variable return flow, for return nodes connected to an airloop with an OA system or not with specified flow (kg/s)
+		Real64 returnSchedFrac = ScheduleManager::GetCurrentScheduleValue( thisZoneEquip.ReturnFlowSchedPtrNum );
```

Comment:
You feel OK calling this without validation of the index variable?  I'm sure it's done elsewhere, so it's probably fine...